editor/gui/content_browser.cpp

```diff
@@ -0,0 +1,183 @@
+#include "content_browser.h"
+
+#include "editor/editor_system.h"
+#include "core/resource_loader.h"
+
+#include "utility/imgui_helpers.h"
+#include "vendor/ImGUI/imgui.h"
+#include "vendor/ImGUI/imgui_impl_dx11.h"
+#include "vendor/ImGUI/imgui_impl_win32.h"
+
+void CALLBACK notifyFileChanges(PVOID lpParam, BOOLEAN TimerOrWaitFired)
+{
+	bool* ReloadPending = (bool*)lpParam;
+	OS::RegisterFileChangesWatcher("game\\assets\\", &notifyFileChanges, ReloadPending);
+	*ReloadPending = true;
+}
+
+void CALLBACK notifyDirectoryChanges(PVOID lpParam, BOOLEAN TimerOrWaitFired)
+{
+	bool* ReloadPending = (bool*)lpParam;
+	OS::RegisterDirectoryChangesWatcher("game\\assets\\", &notifyDirectoryChanges, ReloadPending);
+	*ReloadPending = true;
+}
+
+ContentBrowser::ContentBrowser()
+{
+	m_ReloadPending = true;
+	m_DirectoryImage = ResourceLoader::CreateImageResourceFile("editor\\assets\\icons\\folder.png");
+	m_ScriptImage = ResourceLoader::CreateImageResourceFile("editor\\assets\\icons\\script.png");
+	m_MusicImage = ResourceLoader::CreateImageResourceFile("editor\\assets\\icons\\music.png");
+	OS::RegisterDirectoryChangesWatcher(m_AssetsDirectory, &notifyDirectoryChanges, &m_ReloadPending);
+	OS::RegisterFileChangesWatcher(m_AssetsDirectory, &notifyFileChanges, &m_ReloadPending);
+}
+
+ContentBrowser* ContentBrowser::GetSingleton()
+{
+	static ContentBrowser singleton;
+	return &singleton;
+}
+
+void ContentBrowser::draw(float deltaMilliseconds)
+{
+	ZoneScoped;
+	if (m_ContentBrowserSettings.m_IsActive)
+	{
+		if (ImGui::Begin("Content Browser"))
+		{
+			EditorSystem::GetSingleton()->pushBoldFont();
+			if (m_CurrentDirectory != m_AssetsDirectory)
+			{
+				if (ImGui::Button("<-"))
+				{
+					m_CurrentDirectory = OS::GetParentPath(m_CurrentDirectory).string();
+					m_ReloadPending = true;
+				}
+			}
+			ImGui::SameLine();
+			ImGui::Text(OS::GetRelativePath(m_CurrentDirectory, m_AssetsDirectory).string().c_str());
+			ImGui::SameLine(ImGui::GetWindowWidth() - 85);
+			if (ImGui::Button("Reload"))
+			{
+				m_ReloadPending = true;
+			}
+			EditorSystem::GetSingleton()->popFont();
+			ImGui::Separator();
+
+			int num_columns = ImGui::GetContentRegionAvailWidth() / (m_IconWidth + 30);
+			if (num_columns < 1)
+			{
+				num_columns = 1;
+			}
+			ImGui::Columns(num_columns, 0, false);
+
+			if (m_ReloadPending)
+			{
+				try
+				{
+					Vector<FilePath> filepaths = OS::GetAllInDirectoryRoot(m_CurrentDirectory);
+					Vector<FilePath> temp_files;
+					Vector<FilePath> temp_dirs;
+					m_thumbnail_cache.clear();
+					for (FilePath directoryIterator : filepaths)
+					{
+						String directoryIteratorString = directoryIterator.string();
+						if (OS::IsDirectory(directoryIteratorString))
+						{
+							m_thumbnail_cache[directoryIteratorString] = m_DirectoryImage;
+							temp_dirs.push_back(directoryIteratorString);
+						}
+						else if (directoryIterator.extension().string() == ".wav")
```

Comment:
instead of adding extension string checks everywhere, create a function to get type enum from extension here https://github.com/sdslabs/Rootex/blob/masterv2/rootex/core/resource_loader.h